src/CodeGen_Internal.cpp

```diff
@@ -388,6 +388,42 @@ Expr lower_int_uint_mod(const Expr &a, const Expr &b) {
     }
 }
 
+Expr unsigned_long_div(Expr num, const Expr &den, const Scope<Interval> &bounds) {
+    internal_assert(num.type() == den.type());
+    internal_assert(num.type().is_uint());
+    Type ty = num.type();
+    Expr leading_zeros, q;
+    q = make_zero(ty);
+    leading_zeros = cast(ty, count_leading_zeros(den));
+
+    // Run bounds analysis to estimate the range of result.
+    Expr extent_upper = find_constant_bound(num / den, Direction::Upper, bounds);
+    const uint64_t *upper_value = as_const_uint(extent_upper);
+
+    // Each iteration of the loop below checks for a bit in the result.
+    const int times = ty.bits();
+    for (int i = 1; i <= times; i++) {
+        // Check if the bit at 'shift' index should be set in the result.
+        int shift = times - i;
+        Expr shift_expr = make_const(ty, shift);
+        // Skip iteration for long division if possible.
+        if (upper_value && (*upper_value < (1 << shift))) {
+            debug(1) << "Skipping iteration for long div as "
+                     << "(max_quotient: " << *upper_value << ") < "
+                     << (1 << shift) << ")\n";
+            continue;
```

Comment:
Break? Or modify the loop bound up front instead of this?

Reply:
Done